File: backend/app/schemas/ml.py
"""
Pydantic validation schemas for Machine Learning operations.
"""
from datetime import datetime
from typing import Any, List, Dict, Optional
from uuid import UUID
from pydantic import BaseModel, ConfigDict


class MLModelTrainRequest(BaseModel):
    """Schema representing a request to train a machine learning model."""
    target_column: str
    features: Optional[List[str]] = None
    problem_type: Optional[str] = "auto"
    candidate_algorithms: Optional[List[str]] = None
    cross_validation: bool = False
    random_state: int = 42


class MLModelResponse(BaseModel):
    """Schema representing the details of an MLModel record."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    dataset_id: UUID
    dataset_version_id: UUID
    model_version: int
    target_column: str
    features: List[str]
    problem_type: str
    model_type: str
    status: str
    training_config: Dict[str, Any]
    evaluation_metrics: Optional[Dict[str, Any]] = None
    feature_importances: Optional[Dict[str, float]] = None
    model_size_bytes: Optional[int] = None
    dataset_quality_score: Optional[float] = None
    dataset_hash: Optional[str] = None
    best_score: Optional[float] = None
    training_log: Optional[List[Dict[str, Any]]] = None
    error_message: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    duration_ms: Optional[int] = None
    training_rows: Optional[int] = None
    testing_rows: Optional[int] = None
    created_at: datetime


class PredictionRequest(BaseModel):
    """Schema representing an inference request payload."""
    input_data: Dict[str, Any]


class PredictionResponse(BaseModel):
    """Schema representing the prediction response payload."""
    prediction: Any
    confidence: Optional[float] = None


class PredictionLogResponse(BaseModel):
    """Schema representing an audited prediction history record."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    model_id: UUID
    input_payload: Dict[str, Any]
    prediction: str
    confidence: Optional[float] = None
    created_at: datetime

File: frontend/src/services/ml.ts
/**
 * Frontend API client service for Machine Learning operations.
 */
import { apiClient } from '../lib/apiClient';

export interface MLModelTrainRequest {
  target_column: string;
  problem_type?: string;
  candidate_algorithms?: string[];
  cross_validation?: boolean;
  random_state?: number;
}

export interface MLModelResponse {
  id: string;
  dataset_id: string;
  dataset_version_id: string;
  model_version: number;
  target_column: string;
  features: string[];
  problem_type: string;
  model_type: string;
  status: string;
  training_config: Record<string, any>;
  evaluation_metrics?: Record<string, any>;
  feature_importances?: Record<string, number>;
  model_size_bytes?: number;
  dataset_quality_score?: number;
  dataset_hash?: string;
  best_score?: number;
  training_log?: { timestamp: string; message: string }[];
  error_message?: string;
  started_at?: string;
  completed_at?: string;
  duration_ms?: number;
  training_rows?: number;
  testing_rows?: number;
  created_at: string;
}

export interface PredictionResponse {
  prediction: any;
  confidence?: number;
}

export interface PredictionLogResponse {
  id: string;
  model_id: string;
  input_payload: Record<string, any>;
  prediction: string;
  confidence?: number;
  created_at: string;
}

export const mlService = {
  trainModel: async (datasetId: string, params: MLModelTrainRequest): Promise<any> => {
    const res = await apiClient.post(`/ml/train?dataset_id=${datasetId}`, params);
    return res.data;
  },

  listModels: async (datasetId?: string): Promise<MLModelResponse[]> => {
    const url = datasetId ? `/ml?dataset_id=${datasetId}` : '/ml';
    const res = await apiClient.get(url);
    return res.data.data;
  },

  getModelDetails: async (modelId: string): Promise<MLModelResponse> => {
    const res = await apiClient.get(`/ml/${modelId}`);
    return res.data.data;
  },

  makePrediction: async (modelId: string, inputData: Record<string, any>): Promise<PredictionResponse> => {
    const res = await apiClient.post(`/ml/${modelId}/predict`, { input_data: inputData });
    return res.data.data;
  },

  getPredictionHistory: async (modelId: string): Promise<PredictionLogResponse[]> => {
    const res = await apiClient.get(`/ml/${modelId}/history`);
    return res.data.data;
  },

  deleteModel: async (modelId: string): Promise<void> => {
    await apiClient.delete(`/ml/${modelId}`);
  },
};

export default mlService;
